File: backend/src/backlog_manager_backend/routes/backlog.py
from litestar import Router, delete, get, post, put
from litestar.di import NamedDependency, Provide
from litestar.exceptions import ClientException, NotFoundException, ValidationException
from litestar.params import FromPath, FromQuery
from litestar.status_codes import HTTP_409_CONFLICT
from sqlalchemy.ext.asyncio import AsyncSession

from backlog_manager_backend.auth.dependencies import BEARER_SECURITY_REQUIREMENT, get_current_user
from backlog_manager_backend.errors import ConflictError, NotFoundError
from backlog_manager_backend.repositories import (
    backlog_entry_repo,
    category_backlog_entry_repo,
    category_repo,
    custom_status_repo,
)
from backlog_manager_backend.schemas.backlog_entry import (
    BacklogEntry,
    BacklogEntryResponse,
    CategoryBacklogAssociationParams,
    CreateBacklogEntryParams,
    CreateBacklogEntryRequest,
    GetEntriesByStatusParams,
    UpdateBacklogEntryParams,
    UpdateBacklogEntryRequest,
)
from backlog_manager_backend.schemas.category import (
    Category,
    CategoryResponse,
    CreateCategoryParams,
    CreateCategoryRequest,
    UpdateCategoryParams,
    UpdateCategoryRequest,
)
from backlog_manager_backend.schemas.custom_status import (
    CreateCustomStatusParams,
    CreateCustomStatusRequest,
    CustomStatus,
    CustomStatusResponse,
    UpdateCustomStatusParams,
    UpdateCustomStatusRequest,
)
from backlog_manager_backend.schemas.user import User

_ENTRY_NOT_FOUND = "Backlog entry not found"
_CATEGORY_NOT_FOUND = "Category not found"
_STATUS_NOT_FOUND = "Custom status not found"
DEFAULT_STATUSES = ("Not Started", "In Progress", "Completed", "On Hold", "Dropped")
_STATUS_NAME_MAX_LENGTH = 20


class _StatusConflictException(ClientException):
    """Documents the 409 a duplicate custom-status name raises, distinct
    from ClientException's default 400 so it shows up in the OpenAPI
    response union via the handlers' `raises=` declarations."""

    status_code = HTTP_409_CONFLICT


async def _get_owned_entry(session: AsyncSession, entry_id: int, user: User) -> BacklogEntry:
    """Raises the same response as a real 404 when the entry belongs to
    another user - existence of another user's resource isn't revealed
    by a distinct 403."""
    try:
        entry = await backlog_entry_repo.get_backlog_entry_by_id(session, entry_id)
    except NotFoundError as error:
        raise NotFoundException(_ENTRY_NOT_FOUND) from error
    if entry.user_id != user.id:
        raise NotFoundException(_ENTRY_NOT_FOUND)
    return entry


async def _get_owned_category(session: AsyncSession, category_id: int, user: User) -> Category:
    categories = await category_repo.get_categories_by_user(session, user.id)
    category = next((c for c in categories if c.category_id == category_id), None)
    if category is None:
        raise NotFoundException(_CATEGORY_NOT_FOUND)
    return category


def _join_tags(field_name: str, values: list[str]) -> str:
    """The DB stores genre/platform as one ", "-joined column (unchanged
    from the original schema), so a value containing that delimiter
    would silently split back into multiple values on the next read -
    rejected here rather than persisted lossily."""
    for value in values:
        if "," in value:
            raise ValidationException(f"{field_name} entries must not contain a comma: {value!r}")
    return ", ".join(values)


@post("/api/backlog/entries", status_code=201)
async def create_entry(
    data: CreateBacklogEntryRequest,
    db_session: NamedDependency[AsyncSession],
    current_user: NamedDependency[User],
) -> BacklogEntryResponse:
    entry = await backlog_entry_repo.create_backlog_entry(
        db_session,
        CreateBacklogEntryParams(
            user_id=current_user.id,
            title=data.title,
            genre=_join_tags("genre", data.genre),
            platform=_join_tags("platform", data.platform),
            status=data.status,
            owned=data.owned,
            interest=data.interest,
            release_date=data.release_date,
            image_link=data.image_link,
            main_time=data.main_time,
            main_plus_extra_time=data.main_plus_extra_time,
            completion_time=data.completion_time,
            playtime=data.playtime,
            steam_app_id=data.steam_app_id,
            review_stars=round(data.review_stars) if data.review_stars is not None else None,
            review=data.review,
            note=data.note,
        ),
    )
    return BacklogEntryResponse.from_entry(entry)


@get("/api/backlog/entries")
async def list_entries(
    db_session: NamedDependency[AsyncSession],
    current_user: NamedDependency[User],
    status: FromQuery[str | None] = None,
) -> list[BacklogEntryResponse]:
    if status is not None:
        entries = await backlog_entry_repo.get_backlog_entries_by_status(
            db_session, GetEntriesByStatusParams(user_id=current_user.id, status=status)
        )
    else:
        entries = await backlog_entry_repo.get_backlog_entries_by_user(db_session, current_user.id)
    return [BacklogEntryResponse.from_entry(entry) for entry in entries]


@get("/api/backlog/entries/{entry_id:int}")
async def get_entry(
    entry_id: FromPath[int],
    db_session: NamedDependency[AsyncSession],
    current_user: NamedDependency[User],
) -> BacklogEntryResponse:
    entry = await _get_owned_entry(db_session, entry_id, current_user)
    return BacklogEntryResponse.from_entry(entry)


@put("/api/backlog/entries/{entry_id:int}")
async def update_entry(
    entry_id: FromPath[int],
    data: UpdateBacklogEntryRequest,
    db_session: NamedDependency[AsyncSession],
    current_user: NamedDependency[User],
) -> BacklogEntryResponse:
    await _get_owned_entry(db_session, entry_id, current_user)

    review_stars = data.review_stars
    if isinstance(review_stars, float):
        review_stars = round(review_stars)

    entry = await backlog_entry_repo.update_backlog_entry(
        db_session,
        UpdateBacklogEntryParams(
            backlog_entry_id=entry_id,
            title=data.title,
            genre=_join_tags("genre", data.genre) if isinstance(data.genre, list) else data.genre,
            platform=(
                _join_tags("platform", data.platform)
                if isinstance(data.platform, list)
                else data.platform
            ),
            status=data.status,
            owned=data.owned,
            interest=data.interest,
            release_date=data.release_date,
            image_link=data.image_link,
            main_time=data.main_time,
            main_plus_extra_time=data.main_plus_extra_time,
            completion_time=data.completion_time,
            playtime=data.playtime,
            steam_app_id=data.steam_app_id,
            review_stars=review_stars,
            review=data.review,
            note=data.note,
        ),
    )
    return BacklogEntryResponse.from_entry(entry)


@delete("/api/backlog/entries/{entry_id:int}")
async def delete_entry(
    entry_id: FromPath[int],
    db_session: NamedDependency[AsyncSession],
    current_user: NamedDependency[User],
) -> None:
    await _get_owned_entry(db_session, entry_id, current_user)
    await backlog_entry_repo.delete_backlog_entry(db_session, entry_id)


@delete("/api/backlog/entries", status_code=200)
async def delete_all_entries(
    db_session: NamedDependency[AsyncSession],
    current_user: NamedDependency[User],
) -> int:
    """Bulk variant of delete_entry - permanently deletes every backlog
    entry of the caller (category associations cascade). Returns the
    number of entries removed."""
    return await backlog_entry_repo.delete_backlog_entries_by_user(db_session, current_user.id)


@get("/api/backlog/entries/{entry_id:int}/categories")
async def get_categories_for_entry(
    entry_id: FromPath[int],
    db_session: NamedDependency[AsyncSession],
    current_user: NamedDependency[User],
) -> list[CategoryResponse]:
    await _get_owned_entry(db_session, entry_id, current_user)
    categories = await category_repo.get_categories_for_backlog_entry(db_session, entry_id)
    return [CategoryResponse.from_category(category) for category in categories]


@post("/api/backlog/entries/{entry_id:int}/categories/{category_id:int}", status_code=201)
async def add_category_to_entry(
    entry_id: FromPath[int],
    category_id: FromPath[int],
    db_session: NamedDependency[AsyncSession],
    current_user: NamedDependency[User],
) -> None:
    await _get_owned_entry(db_session, entry_id, current_user)
    await _get_owned_category(db_session, category_id, current_user)
    await category_backlog_entry_repo.add_category_to_backlog_entry(
        db_session,
        CategoryBacklogAssociationParams(category_id=category_id, backlog_entry_id=entry_id),
    )


@delete("/api/backlog/entries/{entry_id:int}/categories/{category_id:int}")
async def remove_category_from_entry(
    entry_id: FromPath[int],
    category_id: FromPath[int],
    db_session: NamedDependency[AsyncSession],
    current_user: NamedDependency[User],
) -> None:
    await _get_owned_entry(db_session, entry_id, current_user)
    await _get_owned_category(db_session, category_id, current_user)
    await category_backlog_entry_repo.remove_backlog_entry_from_category(
        db_session,
        CategoryBacklogAssociationParams(category_id=category_id, backlog_entry_id=entry_id),
    )


@post("/api/backlog/categories", status_code=201)
async def create_category(
    data: CreateCategoryRequest,
    db_session: NamedDependency[AsyncSession],
    current_user: NamedDependency[User],
) -> CategoryResponse:
    category = await category_repo.create_category(
        db_session,
        CreateCategoryParams(
            user_id=current_user.id,
            category_name=data.category_name,
            color=data.color,
            description=data.description,
        ),
    )
    return CategoryResponse.from_category(category)


@get("/api/backlog/categories")
async def list_categories(
    db_session: NamedDependency[AsyncSession],
    current_user: NamedDependency[User],
) -> list[CategoryResponse]:
    categories = await category_repo.get_categories_by_user(db_session, current_user.id)
    return [CategoryResponse.from_category(category) for category in categories]


@put("/api/backlog/categories/{category_id:int}")
async def update_category(
    category_id: FromPath[int],
    data: UpdateCategoryRequest,
    db_session: NamedDependency[AsyncSession],
    current_user: NamedDependency[User],
) -> CategoryResponse:
    await _get_owned_category(db_session, category_id, current_user)
    category = await category_repo.update_category(
        db_session,
        UpdateCategoryParams(
            category_id=category_id,
            category_name=data.category_name,
            color=data.color,
            description=data.description,
        ),
    )
    return CategoryResponse.from_category(category)


@delete("/api/backlog/categories/{category_id:int}")
async def delete_category(
    category_id: FromPath[int],
    db_session: NamedDependency[AsyncSession],
    current_user: NamedDependency[User],
) -> None:
    await _get_owned_category(db_session, category_id, current_user)
    await category_repo.delete_category(db_session, category_id)


@get("/api/backlog/categories/{category_id:int}/entries")
async def get_entries_for_category(
    category_id: FromPath[int],
    db_session: NamedDependency[AsyncSession],
    current_user: NamedDependency[User],
) -> list[BacklogEntryResponse]:
    await _get_owned_category(db_session, category_id, current_user)
    entries = await backlog_entry_repo.get_backlog_entries_for_category(db_session, category_id)
    return [BacklogEntryResponse.from_entry(entry) for entry in entries]


def _validated_status_name(raw_name: str) -> str:
    name = raw_name.strip()
    if not name:
        raise ValidationException("Status name must not be empty")
    if len(name) > _STATUS_NAME_MAX_LENGTH:
        raise ValidationException(
            f"Status name must be at most {_STATUS_NAME_MAX_LENGTH} characters"
        )
    return name


async def _get_owned_status(
    session: AsyncSession, status_id: int, user: User
) -> CustomStatus:
    try:
        status = await custom_status_repo.get_custom_status_by_id(session, status_id)
    except NotFoundError as error:
        raise NotFoundException(_STATUS_NOT_FOUND) from error
    if status.user_id != user.id:
        raise NotFoundException(_STATUS_NOT_FOUND)
    return status


@post(
    "/api/backlog/statuses",
    status_code=201,
    raises=[ValidationException, _StatusConflictException],
)
async def create_custom_status(
    data: CreateCustomStatusRequest,
    db_session: NamedDependency[AsyncSession],
    current_user: NamedDependency[User],
) -> CustomStatusResponse:
    name = _validated_status_name(data.name)
    if name in DEFAULT_STATUSES:
        raise ValidationException(f"{name!r} is already a default status")
    try:
        status = await custom_status_repo.create_custom_status(
            db_session, CreateCustomStatusParams(user_id=current_user.id, name=name)
        )
    except ConflictError as error:
        raise _StatusConflictException(f"A status named {name!r} already exists") from error
    return CustomStatusResponse.from_status(status)


@get("/api/backlog/statuses")
async def list_custom_statuses(
    db_session: NamedDependency[AsyncSession],
    current_user: NamedDependency[User],
) -> list[CustomStatusResponse]:
    statuses = await custom_status_repo.get_custom_statuses_by_user(
        db_session, current_user.id
    )
    return [CustomStatusResponse.from_status(status) for status in statuses]


@put(
    "/api/backlog/statuses/{status_id:int}",
    raises=[NotFoundException, ValidationException, _StatusConflictException],
)
async def update_custom_status(
    status_id: FromPath[int],
    data: UpdateCustomStatusRequest,
    db_session: NamedDependency[AsyncSession],
    current_user: NamedDependency[User],
) -> CustomStatusResponse:
    await _get_owned_status(db_session, status_id, current_user)
    name = _validated_status_name(data.name)
    if name in DEFAULT_STATUSES:
        raise ValidationException(f"{name!r} is already a default status")
    try:
        status = await custom_status_repo.update_custom_status(
            db_session, UpdateCustomStatusParams(status_id=status_id, name=name)
        )
    except ConflictError as error:
        raise _StatusConflictException(f"A status named {name!r} already exists") from error
    return CustomStatusResponse.from_status(status)


@delete("/api/backlog/statuses/{status_id:int}", raises=[NotFoundException])
async def delete_custom_status(
    status_id: FromPath[int],
    db_session: NamedDependency[AsyncSession],
    current_user: NamedDependency[User],
) -> None:
    await _get_owned_status(db_session, status_id, current_user)
    await custom_status_repo.delete_custom_status(db_session, status_id)


backlog_router = Router(
    path="",
    route_handlers=[
        create_entry,
        list_entries,
        get_entry,
        update_entry,
        delete_entry,
        delete_all_entries,
        get_categories_for_entry,
        add_category_to_entry,
        remove_category_from_entry,
        create_category,
        list_categories,
        update_category,
        delete_category,
        get_entries_for_category,
        create_custom_status,
        list_custom_statuses,
        update_custom_status,
        delete_custom_status,
    ],
    dependencies={"current_user": Provide(get_current_user)},
    security=BEARER_SECURITY_REQUIREMENT,
)

File: frontend/src/lib/api/backlog.ts
import { apiClient, apiErrorMessage } from "./client";
import type { components } from "./schema";

export type BacklogStatus =
  | "Not Started"
  | "In Progress"
  | "Completed"
  | "On Hold"
  | "Dropped";

export interface BacklogEntryData {
  id: number;
  title: string;
  imageLink: string;
  imageAlt: string;
  genre: string[];
  platform: string[];
  status: string;
  owned: boolean;
  interest: number;
  reviewStars?: number;
  review?: string;
  note?: string;
  mainTime?: number;
  mainPlusExtraTime?: number;
  completionTime?: number;
  playtime?: number;
  steamAppId?: number;
}

/**
 * `status` is a plain string, not `BacklogStatus`, because the backend
 * doesn't enforce an enum here - CSV imports in particular can carry an
 * arbitrary status value from the source file. `BacklogStatus` stays
 * around for call sites backed by an actual status dropdown
 * (CreationToolForm, BacklogEntry).
 */
export interface CreateBacklogEntryInput {
  title: string;
  genre: string[];
  platform: string[];
  status: string;
  owned: boolean;
  interest: number;
  imageLink?: string;
  mainTime?: number;
  mainPlusExtraTime?: number;
  completionTime?: number;
  playtime?: number;
  steamAppId?: number;
  reviewStars?: number;
  review?: string;
  note?: string;
}

export interface UpdateBacklogEntryInput {
  title?: string;
  genre?: string[];
  platform?: string[];
  status?: string;
  owned?: boolean;
  interest?: number;
  imageLink?: string;
  mainTime?: number;
  mainPlusExtraTime?: number;
  completionTime?: number;
  playtime?: number;
  steamAppId?: number | null;
  reviewStars?: number;
  review?: string;
  note?: string;
}

export interface CategoryData {
  id: number;
  name: string;
  color: string;
  description: string | null;
}

function toNumber(value: string | null | undefined): number | undefined {
  if (value === null || value === undefined) return undefined;
  const parsed = Number(value);
  return Number.isFinite(parsed) ? parsed : undefined;
}

export function toEntryData(
  entry: components["schemas"]["BacklogEntryResponse"],
): BacklogEntryData {
  return {
    id: entry.id,
    title: entry.title,
    imageLink: entry.image_link ?? "",
    imageAlt: entry.title,
    genre: entry.genre,
    platform: entry.platform,
    status: entry.status,
    owned: entry.owned,
    interest: entry.interest,
    reviewStars: entry.review_stars ?? undefined,
    review: entry.review ?? undefined,
    note: entry.note ?? undefined,
    mainTime: toNumber(entry.main_time),
    mainPlusExtraTime: toNumber(entry.main_plus_extra_time),
    completionTime: toNumber(entry.completion_time),
    playtime: toNumber(entry.playtime),
    steamAppId: entry.steam_app_id ?? undefined,
  };
}

function toCategoryData(
  category: components["schemas"]["CategoryResponse"],
): CategoryData {
  return {
    id: category.id,
    name: category.name,
    color: category.color,
    description: category.description ?? null,
  };
}

export async function getEntries(): Promise<BacklogEntryData[]> {
  const { data, error } = await apiClient.GET("/api/backlog/entries");
  if (error)
    throw new Error(apiErrorMessage(error, "Failed to load backlog entries"));
  return data.map(toEntryData);
}

export async function getEntryById(entryId: number): Promise<BacklogEntryData> {
  const { data, error } = await apiClient.GET(
    "/api/backlog/entries/{entry_id}",
    {
      params: { path: { entry_id: entryId } },
    },
  );
  if (error)
    throw new Error(apiErrorMessage(error, "Failed to load backlog entry"));
  return toEntryData(data);
}

export async function getEntriesByStatus(
  status: string,
): Promise<BacklogEntryData[]> {
  const { data, error } = await apiClient.GET("/api/backlog/entries", {
    params: { query: { status } },
  });
  if (error)
    throw new Error(apiErrorMessage(error, "Failed to load backlog entries"));
  return data.map(toEntryData);
}

export async function createEntry(
  input: CreateBacklogEntryInput,
): Promise<BacklogEntryData> {
  const { data, error } = await apiClient.POST("/api/backlog/entries", {
    body: {
      title: input.title,
      genre: input.genre,
      platform: input.platform,
      status: input.status,
      owned: input.owned,
      interest: input.interest,
      image_link: input.imageLink,
      main_time: input.mainTime?.toString(),
      main_plus_extra_time: input.mainPlusExtraTime?.toString(),
      completion_time: input.completionTime?.toString(),
      playtime: input.playtime?.toString(),
      steam_app_id: input.steamAppId,
      review_stars: input.reviewStars,
      review: input.review,
      note: input.note,
    },
  });
  if (error)
    throw new Error(apiErrorMessage(error, "Failed to create backlog entry"));
  return toEntryData(data);
}

export async function updateEntry(
  entryId: number,
  changes: UpdateBacklogEntryInput,
): Promise<BacklogEntryData> {
  const { data, error } = await apiClient.PUT(
    "/api/backlog/entries/{entry_id}",
    {
      params: { path: { entry_id: entryId } },
      body: {
        title: changes.title,
        genre: changes.genre,
        platform: changes.platform,
        status: changes.status,
        owned: changes.owned,
        interest: changes.interest,
        image_link: changes.imageLink,
        main_time: changes.mainTime?.toString(),
        main_plus_extra_time: changes.mainPlusExtraTime?.toString(),
        completion_time: changes.completionTime?.toString(),
        playtime: changes.playtime?.toString(),
        steam_app_id: changes.steamAppId,
        review_stars: changes.reviewStars,
        review: changes.review,
        note: changes.note,
      },
    },
  );
  if (error)
    throw new Error(apiErrorMessage(error, "Failed to update backlog entry"));
  return toEntryData(data);
}

export async function deleteEntry(entryId: number): Promise<void> {
  const { error } = await apiClient.DELETE("/api/backlog/entries/{entry_id}", {
    params: { path: { entry_id: entryId } },
  });
  if (error)
    throw new Error(apiErrorMessage(error, "Failed to delete backlog entry"));
}

export async function getCategories(): Promise<CategoryData[]> {
  const { data, error } = await apiClient.GET("/api/backlog/categories");
  if (error)
    throw new Error(apiErrorMessage(error, "Failed to load categories"));
  return data.map(toCategoryData);
}

export async function createCategory(input: {
  categoryName: string;
  color?: string;
  description?: string;
}): Promise<CategoryData> {
  const { data, error } = await apiClient.POST("/api/backlog/categories", {
    body: {
      category_name: input.categoryName,
      color: input.color,
      description: input.description,
    },
  });
  if (error)
    throw new Error(apiErrorMessage(error, "Failed to create category"));
  return toCategoryData(data);
}

export async function updateCategory(
  categoryId: number,
  changes: { categoryName?: string; color?: string; description?: string },
): Promise<CategoryData> {
  const { data, error } = await apiClient.PUT(
    "/api/backlog/categories/{category_id}",
    {
      params: { path: { category_id: categoryId } },
      body: {
        category_name: changes.categoryName,
        color: changes.color,
        description: changes.description,
      },
    },
  );
  if (error)
    throw new Error(apiErrorMessage(error, "Failed to update category"));
  return toCategoryData(data);
}

export async function deleteCategory(categoryId: number): Promise<void> {
  const { error } = await apiClient.DELETE(
    "/api/backlog/categories/{category_id}",
    {
      params: { path: { category_id: categoryId } },
    },
  );
  if (error)
    throw new Error(apiErrorMessage(error, "Failed to delete category"));
}

export async function getCategoriesForEntry(
  entryId: number,
): Promise<CategoryData[]> {
  const { data, error } = await apiClient.GET(
    "/api/backlog/entries/{entry_id}/categories",
    { params: { path: { entry_id: entryId } } },
  );
  if (error)
    throw new Error(
      apiErrorMessage(error, "Failed to load categories for entry"),
    );
  return data.map(toCategoryData);
}

export async function getEntriesForCategory(
  categoryId: number,
): Promise<BacklogEntryData[]> {
  const { data, error } = await apiClient.GET(
    "/api/backlog/categories/{category_id}/entries",
    { params: { path: { category_id: categoryId } } },
  );
  if (error)
    throw new Error(
      apiErrorMessage(error, "Failed to load entries for category"),
    );
  return data.map(toEntryData);
}

export async function addCategoryToEntry(
  entryId: number,
  categoryId: number,
): Promise<void> {
  const { error } = await apiClient.POST(
    "/api/backlog/entries/{entry_id}/categories/{category_id}",
    { params: { path: { entry_id: entryId, category_id: categoryId } } },
  );
  if (error)
    throw new Error(apiErrorMessage(error, "Failed to add category to entry"));
}

export async function removeCategoryFromEntry(
  entryId: number,
  categoryId: number,
): Promise<void> {
  const { error } = await apiClient.DELETE(
    "/api/backlog/entries/{entry_id}/categories/{category_id}",
    { params: { path: { entry_id: entryId, category_id: categoryId } } },
  );
  if (error)
    throw new Error(
      apiErrorMessage(error, "Failed to remove category from entry"),
    );
}
